### PR TITLE
Platform Tags

### LINKED ISSUE #643: SpongeForge/SpongeVanilla Tags
If a project has a SpongeForge / SpongeVanilla dependency, we should display that _instead of_ the Sponge / Forge dependencies it may have.

### PR DESCRIPTION
**Pull request by phase**
Add SpongeForge, SpongeVanilla, SpongeCommon, and Lantern tags. Sponge & Forge tags won't be shown if an SF/SV dependency is present. Read the comments below for more details.

Closes #643

### CODE REVIEW

File: app/ore/project/Dependency.scala [hunk 1]
```diff
@@ -40,4 +40,8 @@ object Dependency {
     */
   val ForgeId: String = "forge"
 
+  val SpongeForgeId: String = "spongeforge"
```
Comment (stephan-gh) [inline app/ore/project/Dependency.scala]: Problem: There is no `spongeforge`/`spongevanilla` plugin ID. Both implementations use `sponge` as plugin ID currently.

**Comment by phase** [comment]
IDs:
* SpongeAPI - `spongeapi`
* SpongeCommon - `sponge`
* SpongeForge - `spongeforge`
* SpongeVanilla - `spongevanilla`

These will be getting Plugin Containers soon(tm).

**Comment by XakepSDK** [comment]
@phase 
Why SpongeCommon - `sponge`?

**Comment by phase** [comment]
Sponge uses `sponge` for a lot of things internally and we want something to have that mod id.

**Comment by XakepSDK** [comment]
Sponge may have multiple implementations, saying that SpongeCommon is `sponge` is wrong.

**Comment by Zidane** [comment]
@XakepSDK 

Doesn't matter, my implementation in common is always known as `sponge`.

All other ones are known differently. SpongeForge and SpongeVanilla will gain a container with their own id.

**Comment by phase** [comment]
Here's a preview of a project creation with the new platform:
2018-08-16-142248_1920x1080_scrot

And here's a version list with different variations:

2018-08-16-142308_1920x1080_scrot

The dependencies for each version:
* **1.0.4**: `forge@blah, spongeforge@blah, spongeapi@blah`
* **1.0.3**: `forge@blah, spongeapi@blah`
* **1.0.2**: `spongeforge@blah, spongeapi@blah`
* ~~1.0.1: ignore that one I messed up~~
* **1.0.0**: `spongeapi@blah`

SpongeCommon isn't tracked but I guess it could if we wanted it to be.
I rewrote the Platform system so that adding a new one is _super_ simple.

I'm also open to suggestions for the **COLORS OF TAGS!!!!!**, the most important part. Currently, SpongeForge and SpongeVanilla (which also works with this system) use the Sponge tag, but we can use anything.

**Comment by phase** [comment]
colors!

SF & SV will both be displayed if they are both depended on (Platforms with the same priority are shown together).

**Comment by phase** [comment]
Colors have been changed!



And if a project's recommended version has more than 2 tags, the data will be excluded on the home page (this screenshot doesn't have the new colors).

**Comment by phase** [comment]
working on a platform filter now

File: app/controllers/Application.scala [hunk 2]
```diff
@@ -87,17 +85,21 @@ final class Application @Inject()(data: DataHelper,
...
 
     val ordering = sort.flatMap(ProjectSortingStrategies.withId).getOrElse(ProjectSortingStrategies.Default)
-    // TODO platform filter is not implemented
+    val pcat = platformCategory.flatMap(p => PlatformCategory.getPlatformCategories.find(_.name.equalsIgnoreCase(p)))
     val pform = platform.flatMap(p => Platforms.values.find(_.name.equalsIgnoreCase(p)).map(_.asInstanceOf[Platform]))
-    // val platformFilter = pform.map(actions.platformFilter).getOrElse(ModelFilter.Empty)
+
+    // get the categories being queried
+    val categoryPlatformNames: List[String] = pcat.toList.flatMap(_.getPlatforms.map(_.name))
+    val platformNames: List[String] = pform.map(_.name).toList.:::(categoryPlatformNames).map(_.toLowerCase)
```
Comment (Katrix) [inline app/controllers/Application.scala]: `::: ` should be called infix

File: app/models/project/Project.scala [hunk 3]
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
     if (isDefined) update(RecommendedVersionId) else Future.unit
   }
 
+  /**
+    * Get a collection of tags that represent a project through its versions
+    */
+  def tags(implicit ec: ExecutionContext, service: ModelService = null): Future[List[Tag]] = {
```
Comment (Katrix) [inline app/models/project/Project.scala]: Can we please try to get away from this pattern with an implicit value with a default null?

File: app/models/project/Project.scala [hunk 4]
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
...
+    schema(service)
+    // get all the versions for the project
+    this.service.access(classOf[Version]).filter(_.projectId === id.get).flatMap { versions =>
+      val tagIds = versions.flatMap(_.tagIds).toSet.toList
+      // get all the tags for all the versions
+      this.service.access(classOf[Tag]).filter(t => t.id inSet tagIds).map { list =>
+        list.toSet.toList
```
Comment (Katrix) [inline app/models/project/Project.scala]: There's a method for this called `distinct`

File: app/models/project/Project.scala [hunk 5]
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
...
+          // get the latest tag from the versions
+          .groupBy[String](_.name)
```
Comment (Katrix) [inline app/models/project/Project.scala]: Is the type parameter needed here?

File: app/models/project/Project.scala [hunk 6]
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
...
+          .map(pair => {
```
Comment (Katrix) [inline app/models/project/Project.scala]: Only use curly braces here. Also, name the pair's elements with a `case`

File: app/models/project/Project.scala [hunk 7]
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
...
+            pair._2.maxBy(
+              t => versions
+                .filter(_.tagIds.contains(t.id.get))
+                .filter(!_.isDeleted)
+                // get the latest version
+                .map(_.createdAt.getOrElse(Timestamp.from(Instant.MIN)).toInstant.toEpochMilli)
```
Comment (Katrix) [inline app/models/project/Project.scala]: Don't be afraid to get here. It's an error if `createdAt` is empty.

File: app/models/project/Tag.scala [hunk 8]
```diff
@@ -23,22 +26,47 @@ case class Tag(override val id: Option[Int] = None,
...
 
   def copyWith(id: Option[Int], theTime: Option[Timestamp]): Tag = this.copy(id = id)
+
+  /**
+    * Used to convert a ghost tag to a normal tag
+    * @author phase
+    */
+  def getFilledTag(service: ModelService)(implicit ex: ExecutionContext): Future[Tag] = {
+    val access = service.access(classOf[Tag])
+    for {
```
Comment (Katrix) [inline app/models/project/Tag.scala:44]: Just a styling preference, can we get rid of the curly braces for tag and yield here?

File: app/ore/Platforms.scala [hunk 9]
```diff
@@ -1,13 +1,86 @@
...
+    def toGhostTag(version: String): Tag = Tag(None, List(), name, version, tagColor)
+
+  }
+
+  implicit def convert(v: Value): Platform = v.asInstanceOf[Platform]
+
+  def getPlatforms(dependencyIds: List[String]): List[Platform] = {
+    Platforms.values
+      .filter(p => dependencyIds.contains(p.dependencyId))
+      .groupBy[PlatformCategory](p => p.platformCategory)
```
Comment (Katrix) [inline app/ore/Platforms.scala]: Is the type parameter needed here?

File: app/ore/Platforms.scala [hunk 10]
```diff
@@ -1,13 +1,86 @@
...
+      .flatMap(map => map._2.groupBy(p => p.priority).maxBy(_._1)._2)
+      .map(p => p.asInstanceOf[Platform])
+      .toList
+  }
+
+  def getPlatformGhostTags(dependencies: List[Dependency]): List[Tag] = {
+    Platforms.values
+      .filter(p => dependencies.map(_.pluginId).contains(p.dependencyId))
+      .groupBy[PlatformCategory](p => p.platformCategory)
```
Comment (Katrix) [inline app/ore/Platforms.scala]: Is the type parameter needed here?

File: app/ore/project/factory/ProjectFactory.scala [hunk 11]
```diff
@@ -374,44 +370,23 @@ trait ProjectFactory {
...
-          }
-        } yield {
-          newVersion.addTag(tag)
-          tag
-        }
+  private def addDependencyTags(version: Version)(implicit ec: ExecutionContext): Future[Seq[ProjectTag]] = {
+    for {
+      tags <- Future.sequence(Platforms.getPlatformGhostTags(version.dependencies).map(tag => tag.getFilledTag(service)))
+    } yield {
+      tags.map(tag => {
```
Comment (Katrix) [inline app/ore/project/factory/ProjectFactory.scala]: Only curly braces here

File: app/models/project/Project.scala (hunk 3 above)
Comment (phase) [reply app/models/project/Project.scala]: I just copied the code from `Version#tags`

File: app/models/project/Project.scala [hunk 12]
```diff
@@ -507,27 +507,27 @@ case class Project(override val id: Option[Int] = None,
...
+        list.distinct
           // get the latest tag from the versions
-          .groupBy[String](_.name)
-          .map(pair => {
-            pair._2.maxBy(
-              t => versions
-                .filter(_.tagIds.contains(t.id.get))
+          .groupBy(_.name)
+          .map { case (_, tags) =>
+            tags.maxBy(
```
Comment (Katrix) [inline app/models/project/Project.scala]: Curly braces here

**Comment by phase** [comment]
I updated the Dependencies section for versions. They now include links to the download pages for whatever the platform is.

2018-08-18-145748_1920x1080_scrot

Here's a final view of the hompage being filtered by forge mods.

2018-08-18-145606_1920x1080_scrot

File: app/models/project/Project.scala [hunk 13]
```diff
@@ -504,6 +504,34 @@ case class Project(override val id: Option[Int] = None,
...
   }
 
+  /**
+    * Get a collection of tags that represent a project through its versions
+    */
+  def tags(implicit ec: ExecutionContext, service: ModelService): Future[List[Tag]] = {
+    schema(service)
+    // get all the versions for the project
+    this.service.access(classOf[Version]).filter(_.projectId === id.get).flatMap { versions =>
+      val tagIds = versions.flatMap(_.tagIds).toSet.toList
```
Comment (Katrix) [inline app/models/project/Project.scala]: Use distinct here

File: app/ore/project/factory/ProjectFactory.scala [hunk 14]
```diff
@@ -374,44 +370,22 @@ trait ProjectFactory {
...
-              tag.addVersionId(newVersion.id.get)
-              Future.successful(tag)
-            }
-          }
-        } yield {
-          newVersion.addTag(tag)
-          tag
-        }
+  private def addDependencyTags(version: Version)(implicit ec: ExecutionContext): Future[Seq[ProjectTag]] = {
+    for {
```
Comment (Katrix) [inline app/ore/project/factory/ProjectFactory.scala]: Do we need the for comprehension here?

**Review by Ewoutvans — APPROVED** [review]

File: app/ore/Platforms.scala [hunk 15]
```diff
@@ -1,13 +1,98 @@
...
+
+  implicit def convert(v: Value): Platform = v.asInstanceOf[Platform]
+
+  def getPlatforms(dependencyIds: Seq[String]): Seq[Platform] = {
+    Platforms.values
+      .filter(p => dependencyIds.contains(p.dependencyId))
+      .groupBy(_.platformCategory)
+      .flatMap(_._2.groupBy(_.priority).maxBy(_._1)._2)
+      .map(_.asInstanceOf[Platform])
+      .toList
```
Comment (Katrix) [inline app/ore/Platforms.scala]: Be explicit about the return type if you're explicit about the Seq implementation.

File: app/ore/Platforms.scala [hunk 16]
```diff
@@ -1,13 +1,98 @@
...
+  }
+
+  def getPlatformGhostTags(dependencies: Seq[Dependency]): Seq[Tag] = {
+    Platforms.values
+      .filter(p => dependencies.map(_.pluginId).contains(p.dependencyId))
+      .groupBy(_.platformCategory)
+      .flatMap(_._2.groupBy(_.priority).maxBy(_._1)._2)
+      .map(p => p.toGhostTag(dependencies.find(_.pluginId == p.dependencyId).get.version))
+      .toList
```
Comment (Katrix) [inline app/ore/Platforms.scala]: Be explicit about the return type if you're explicit about the Seq implementation.

File: app/ore/Platforms.scala [hunk 17]
```diff
@@ -1,13 +1,98 @@
...
+  * Bukkit <- Bukkit, Spigot, Paper
+  * Canary <- Canary, Neptune
+  *
+  * @author phase
+  */
+sealed trait PlatformCategory {
+  val name: String
+
+  def getPlatforms: Seq[Platforms.Value] = {
+    Platforms.values.filter(p => p.platformCategory == this).toList
```
Comment (Katrix) [inline app/ore/Platforms.scala]: Be explicit about the return type if you're explicit about the Seq implementation.

**Review by Katrix — APPROVED** [review]

**Comment by phase** [comment]
I changed snyk not to check OreTestPlugin on PRs.

**Pull request merged**